### DISCUSSION TITLE
Using `oneofs=union`, how would I get a list of the oneof types?

### DISCUSSION THREAD

**Discussion by natebot13** [stephenh/ts-proto#894] [Q&A]
I'm able to get a type union of all the oneofs from a proto using something like this:
```ts
type MyProtoTypes = NonNullable<MyProtoWithOneofs['OptionType']>['$case'];
```
But it seems that typescript doesn't support the ability to then turn that type union into a list that I can use at runtime. Is there a way to get the list of oneof types as strings? Perhaps ts-proto could generate enums for the oneof types, like other proto compilers?

**Comment by stephenh** [comment]
Hi @natebot13 , no we do not have that atm. We could potentially add the lists to the message const, i.e. here:

https://github.com/stephenh/ts-proto/blob/main/integration/oneof-unions/oneof.ts#L80

Have something like:

```
export const PleaseChoose = {
  choices = ["aNumber", "aString", ...];
  eitherOrs = ["either", "or", "thirdOption"];
```

Granted, I'm kind of assuming you'd need more than just the names, like maybe this a map of the `$case` name to something...

We do have a `outputTypeRegistry` option, which doesn't do this today, but sounds similar-ish, in terms of "encode the metadata into runtime data structures that I can introspect", so if we add anything for `$case` lists, we could put it behind that flag.

**Comment by natebot13** [comment]
Got it, yeah adding the list of oneof types to the message const makes sense. I'm not sure yet about what extra data would be useful to go along with that. 
I'm currently just creating a const object with keys of the type union, which forces me to ensure all the fields are specified, and then I can use Object.keys on that object. Something like this:
```typescript
type MyProtoType = NonNullable<MyProtoWithOneofs['OptionType']>['$case'];
type ProtoTypeTemplate = { [key of MyProtoType]: boolean }
const ProtoTypeObject: ProtoTypeTemplate {
  option1: true,
 ... // Forced to add all the types here by the compiler
}
const MY_PROTO_TYPES = Object.keys(ProtoTypeObject).map(s => s as MyProtoType)
```
